sdk/samples/iot/iot_sample_common.h

```diff
@@ -0,0 +1,190 @@
+
+char iot_sample_hub_hostname_buffer[128];
+char iot_sample_provisioning_id_scope_buffer[16];
+
+char iot_sample_hub_device_id_buffer[64];
+char iot_sample_provisioning_registration_id_buffer[256];
+
+char iot_sample_hub_sas_key_buffer[128];
+char iot_sample_provisioning_sas_key_buffer[128];
+
+char iot_sample_x509_cert_pem_file_path_buffer[256];
+char iot_sample_x509_trust_pem_file_path_buffer[256];
+
+typedef struct
+{
+  az_span hub_hostname;
+  az_span provisioning_id_scope;
+  az_span hub_device_id;
+  az_span provisioning_registration_id;
+  az_span hub_sas_key;
+  az_span provisioning_sas_key;
+  uint32_t sas_key_duration_minutes;
+  az_span x509_cert_pem_file_path;
+  az_span x509_trust_pem_file_path;
+} iot_sample_environment_variables;
+
+typedef enum
+{
+  PAHO_IOT_HUB,
+  PAHO_IOT_PROVISIONING
+} iot_sample_type;
+
+typedef enum
+{
+  PAHO_IOT_HUB_C2D_SAMPLE,
+  PAHO_IOT_HUB_METHODS_SAMPLE,
+  PAHO_IOT_HUB_PNP_COMPONENT_SAMPLE,
+  PAHO_IOT_HUB_PNP_SAMPLE,
+  PAHO_IOT_HUB_SAS_TELEMETRY_SAMPLE,
+  PAHO_IOT_HUB_TELEMETRY_SAMPLE,
+  PAHO_IOT_HUB_TWIN_SAMPLE,
+  PAHO_IOT_PROVISIONING_SAMPLE,
+  PAHO_IOT_PROVISIONING_SAS_SAMPLE
+} iot_sample_name;
+
+/*
+ * @brief Reads in environment variables set by user for purposes of running sample.
+ *
+ * @param[in] type The enumerated type of the sample.
+ * @param[in] name The enumerated name of the sample.
+ * @param[out] out_env_vars A pointer to the struct containing all read-in environment variables.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ * @retval #AZ_OK All required environment variables successfuly read-in.
+ * @retval #AZ_ERROR_ARG Sample type or name is undefined, or environment variable is not set.
+ * @retval #AZ_RETURN_IF_NOT_ENOUGH_SIZE Not enough space set aside to store environment variable.
+ */
+az_result iot_sample_read_environment_variables(
+    iot_sample_type type,
+    iot_sample_name name,
+    iot_sample_environment_variables* out_env_vars);
+
+/*
+ * @brief Builds an MQTT endpoint c-string for an Azure IoT Hub or provisioning service.
+ *
+ * @param[in] type The enumerated type of the sample.
+ * @param[in] env_vars A pointer to environment variable struct.
+ * @param[out] out_endpoint A pointer to char buffer containing the built c-string.
+ * @param[in] endpoint_size The size of the char buffer to be filled.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ * @retval #AZ_OK MQTT endpoint successfully created.
+ * @retval #AZ_ERROR_ARG Sample type is undefined.
+ * @retval #AZ_ERROR_INSUFFICIENT_SPAN_SIZE Buffer size is not large enough to hold c-string.
+ */
+az_result iot_sample_create_mqtt_endpoint(
+    iot_sample_type type,
+    iot_sample_environment_variables const* env_vars,
+    char* out_endpoint,
+    size_t endpoint_size);
+
+/*
+ * @brief Sleep for given seconds.
+ *
+ * @param[in] seconds Number of seconds to sleep.

```

Comment:
@antkmsft - I don't know how helpful `[in]` is for primitive value types being passed by value (rather than as a reference using `*`). Isn't it implicit/given that such parameters are `[in]`. We _could_ just leave these unspecified. What do you think?

Reply:
While it is likely obvious to the developer that it is an `in`, I do like being consistent with the [in] on `in`s.  Then there is no question about what we mean when there is nothing mentioned (like in the case of an az_span that will be changed but not used, e.g. destination).

I do see [in] it used for primitive types in a few functions here: https://github.com/Azure/azure-sdk-for-c/blob/f36af1eb3c47b96ecf1f5c4010efb3d9e6ea261f/sdk/inc/azure/core/az_json.h#L453

Reply:
> While it is likely obvious to the developer that it is an `in`, I do like being consistent with the [in] on `in`s.

Sure, makes sense. The contribution to "noise" in the docs would be minimal anyway.

> I do see [in] used for primitive types in a few functions here:

Yes, I saw those too, which is why my question isn't about your particular PR or code change (what you have is good). It just reminded me about it. :)

We have some rare instances in the repo that don't use `[in]`:
https://github.com/Azure/azure-sdk-for-c/blob/730e2c3ed7a2fb8952b5e8b6ba5b0abbc2ff5019/sdk/src/azure/core/az_span_private.h#L27-L35

Reply:
ah interesting.  This would be a good item for us to agree upon in the next C99 meeting.  I am finding some other issues like this.  What if I start an email listing the issues (code vs. guideline differences, or code vs no guideline spec).  We can gather what we are finding before shipping it off to a C99 meeting?

Reply:
Great idea. Go for it.